lib/bonus.stk

```diff
@@ -1124,6 +1127,109 @@ doc>
       (begin ,@body
              (-> #:top)))))
 
+#|
+<doc EXT 1+ 1-
+ * (1+ x)
+ * (1- x)
+ *
+ * There procedures return |x| plus one and |x| minus one, repsectively.
+ * Being procedures, they do not change the value of the variable |x|.
+ *
+ * @lisp
+ * (define x 10)
+ * (1+ x)          => 11
+ * x               => 10
+ * (1- x)          => 9
+ * x               => 10
+ * @end lisp
+doc>
+|#
+(define (1+ x) (+ x 1))
+(define (1- x) (- x 1))
+
+
+#|
+<doc EXT inc! dec!
+ * (inc! x)
+ * (dec! x)
+ *
+ * There macros change the value of |x| to |x|+1 and |x|-1, respectively.
+ *
+ * @lisp
+ * (define x 10)
+ * (inc! x)
+ * x               => 11
+ * (dec! x)
+ * x               => 10
+ * @end lisp
+doc>
+|#
+(define-macro (inc! x)
+  `(set! ,x (+ ,x 1)))
+
+(define-macro (dec! x)
+  `(set! ,x (- ,x 1)))
+
+
+#|
+<doc EXT push! pop!
+ * (push! x lst)
+ * (pop! lst)
+ *
+ * These macros push and pop an element into/from a list, so as to use
+ * it as a stack.
+ *
+ * @lisp
+ * (define S (list 20 30 40))
+ * (push! 10 S)
+ * S                         => (10 20 30 40)
+ * (pop! S)                  => 10
+ * S                         => (20 30 40)
+ * @end lisp
+doc>
+|#
+(define-macro (push! e stack)
+  `(set! ,stack (cons ,e ,stack)))
+
+(define-macro (pop! stack)
+  (let ((x (gensym 'res-)))
+    `(let ((,x (car ,stack)))
+       (set! ,stack (cdr ,stack))
+       ,x)))
+
+
+#|
+<doc EXT dolist
+ * (dolist (x lst) body)
+ *
```

Comment:
Common Lisp has an optional result form which you may or may not want to have in STklos: `(dolist (x lst [result]) body)`. Same with `dotimes`.

Reply:
Yes - but since STklos' `dotimes` (and Gauche's) do not return a result, I thought this would keep the system consistent.